src/main/webapp/feed.html

```diff
@@ -22,72 +25,67 @@ <h1>Message Feed</h1>
-
+      const [messagesByUser, setMessages] = React.useState(["No messages!", h('br')])
       React.useEffect(fetchMessages, []);
 
       // Fetch messages and add them to the page.
       function fetchMessages() {
-        // store message react components
-        const fetchedMessages = []
-        // store how many messages each user posts in map for later use
-        const userMessageCounts = {}
-        // store react components of the num of msgs each user posts
-        const finalUserMessageCounts = ["Number of Messages Per User:", h('br')]
+        // Store message React components in JS Map with emails as the key
+        const fetchedMessagesByUser = new Map();
+        const completeMessageDiv = [];
         const url = '/feed';
         fetch(url)
           .then(response => response.json())
           .then(messages => {
-            // get data from each message
+            // Map messages to the users who posted them
             messages.forEach(message => {
-              // if a user has posted before, incr
-              if (message.user in userMessageCounts) userMessageCounts[message.user]++
-              // otherwise set their msg counts to 1
-              else userMessageCounts[message.user] = 1
-              // push message react component to list
-              fetchedMessages.push(buildMessageDiv(message));
+              const {user} = message;
+              const newMessage = buildMessageDiv(message);
+              if (!fetchedMessagesByUser.has(user))
+                fetchedMessagesByUser.set(user, [newMessage]);
+              else
+                fetchedMessagesByUser.set(
+                  user, [...fetchedMessagesByUser.get(user), newMessage]
+                );
             });
           })
           .then(() => {
-            // update render state of all messages posted
-            setMessages(fetchedMessages);
-            // go thru each user and their counts and push a react component to the list
-            for(user in userMessageCounts)
-              finalUserMessageCounts.push(buildUserCountDiv(user, userMessageCounts[user]))
-          })
-          .then(() => {
-            // update render state of each users message counts if msgs pushed
-            if (finalUserMessageCounts > 2)
-              setNumUserMessages(finalUserMessageCounts);
+            if (fetchedMessagesByUser.size) {
+              // After adding all messages, append message counts (length of message array)
+              // to the beginning of each user's message array *if messages were added*
+              fetchedMessagesByUser.forEach((userMessages, user) => {
+                completeMessageDiv.push(
+                  [buildNumUserMessagesDiv(user, userMessages.length), ...userMessages]
+                );
+              })
+              setMessages(completeMessageDiv.map((div, index) => h('div', {key: index}, div)));
```

Comment:
Again, you're having trouble because you're trying to store opaque (read: non-introspectable elements) inside state. Generally, people avoid storing React elements in state.